src/Stratis.Bitcoin.Features.Consensus.Tests/TimeSyncBehaviorTest.cs

```diff
@@ -28,173 +28,283 @@ public class TimeSyncBehaviorTest
-                allSamples.AddRange(outSamples);
+            // add enough outbound samples to get an adjustment - and across an even range
+            Assert.True(state.AddTimeData( new IPAddress(1), TimeSpan.FromSeconds(-500), false));
+            Assert.True(state.AddTimeData( new IPAddress(2), TimeSpan.FromSeconds(-250), false));
+            Assert.True(state.AddTimeData( new IPAddress(3), TimeSpan.FromSeconds(250), false));
+            Assert.True(state.AddTimeData( new IPAddress(4), TimeSpan.FromSeconds(500), false));
+
+            TimeSpan diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - 250*1000) < TimeEpsilonMs, $"should be 250ms because leaning towards the inbound value of 400ms, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+
+            // add another batch of inbound with a greater offset than highest outbound
+            for (int i = TimeSyncBehaviorState.MaxInboundSamples; i < TimeSyncBehaviorState.MaxInboundSamples*2; i++)
+            {
+                Assert.True(state.AddTimeData(new IPAddress(i*2), TimeSpan.FromSeconds(-1000), true), $"index: {i}");
+            }
+
+            diff =  dateTimeProvider.GetAdjustedTime() -  dateTimeProvider.GetUtcNow();
+            Assert.True(Math.Abs(diff.TotalMilliseconds - (-250*1000)) < TimeEpsilonMs, $"should be -250ms because the new inbound replace the old and swing the median closer to the new -1000ms values, but still in outbound. Actual: {diff.TotalMilliseconds}ms");
+        }
+         
+        /// <summary>
+        /// Sets up the system under test
+        /// </summary>
+        /// <param name="dateTimeProvider"></param>
+        /// <returns></returns>
+        private TimeSyncBehaviorState CreateTimeSyncBehaviorState(DateTimeProvider dateTimeProvider)
+        {
+            var lifetime = new NodeLifetime();
+            var loggerFactory = new LoggerFactory();
+            var asyncLoopFactory = new AsyncLoopFactory(loggerFactory);
+            var state = new TimeSyncBehaviorState(dateTimeProvider, lifetime, asyncLoopFactory, loggerFactory, Network.Main);
+            return state;
+        }
+    }
+
+    /// <summary>
+    /// Representation of a Sample for data driven tests.
+    /// </summary>
+    public class TestSample
+    {
+        /// <summary>
+        /// Input isInbound flag, with false meaning outbound.
+        /// </summary>
+        public bool IsInbound { get; }
 
-            adjustedTime = dateTimeProvider.GetAdjustedTime();
-            normalTime = dateTimeProvider.GetUtcNow();
-            diff = adjustedTime - normalTime;
-            expectedDiffMs = allSamples.Median();
-            Assert.True(Math.Abs(diff.TotalMilliseconds) < Math.Abs(expectedDiffMs) + TimeEpsilonMs);
+        /// <summary>
+        /// Expectation of whether the sample is used or not.
```

Comment:
You could add information in which cases we expect the sample not to be used

Reply:
I could